File: examples/java/cases/tasks/task_create/TaskCreateDomainCase.java
package cases.tasks.task_create;

import core.BaseDomainCase;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class TaskCreateDomainCase
    extends BaseDomainCase<TaskCreateDomainCase.TaskCreateInput, TaskCreateDomainCase.TaskCreateOutput> {
  public static final List<String> TASK_STATUS_VALUES = List.of("todo", "doing", "done");

  public record Task(
      String id,
      String title,
      String description,
      String status,
      String createdAt,
      String updatedAt
  ) {
    public static Task fromMap(Map<String, Object> value, String source) {
      if (value == null) {
        throw new IllegalArgumentException(source + " must be an object");
      }

      String id = requiredString(value, "id", source);
      String title = requiredString(value, "title", source);
      String status = requiredString(value, "status", source);
      String createdAt = requiredString(value, "createdAt", source);
      String updatedAt = requiredString(value, "updatedAt", source);
      Object description = value.get("description");

      if (!TASK_STATUS_VALUES.contains(status)) {
        throw new IllegalArgumentException(
            source + ".status must be one of " + String.join(", ", TASK_STATUS_VALUES)
        );
      }

      if (description != null && !(description instanceof String)) {
        throw new IllegalArgumentException(source + ".description must be a string when provided");
      }

      return new Task(id, title, (String) description, status, createdAt, updatedAt);
    }

    public Map<String, Object> toMap() {
      Map<String, Object> value = new LinkedHashMap<>();
      value.put("id", id);
      value.put("title", title);
      if (description != null && !description.isBlank()) {
        value.put("description", description);
      }
      value.put("status", status);
      value.put("createdAt", createdAt);
      value.put("updatedAt", updatedAt);
      return value;
    }
  }

  public record TaskCreateInput(String title, String description) {
    public static TaskCreateInput fromMap(Map<String, Object> value) {
      if (value == null) {
        throw new IllegalArgumentException("input must be an object");
      }

      for (String forbidden : List.of("id", "status", "createdAt", "updatedAt")) {
        if (value.containsKey(forbidden)) {
          throw new IllegalArgumentException(forbidden + " must not be provided by the caller");
        }
      }

      for (String key : value.keySet()) {
        if (!List.of("title", "description").contains(key)) {
          throw new IllegalArgumentException("Unexpected input field " + key);
        }
      }

      Object title = value.get("title");
      Object description = value.get("description");

      if (!(title instanceof String)) {
        throw new IllegalArgumentException("title is required and must be a string");
      }

      if (description != null && !(description instanceof String)) {
        throw new IllegalArgumentException("description must be a string when provided");
      }

      return new TaskCreateInput((String) title, (String) description);
    }
  }

  public record TaskCreateOutput(Task task) {
    public static TaskCreateOutput fromMap(Map<String, Object> value) {
      if (value == null || !(value.get("task") instanceof Map<?, ?> taskMap)) {
        throw new IllegalArgumentException("task_create output must contain task");
      }

      @SuppressWarnings("unchecked")
      Map<String, Object> taskValue = (Map<String, Object>) taskMap;
      return new TaskCreateOutput(Task.fromMap(taskValue, "task_create.output.task"));
    }

    public Map<String, Object> toMap() {
      return Map.of("task", task.toMap());
    }
  }

  @Override
  public String caseName() {
    return "task_create";
  }

  @Override
  public String description() {
    return "Creates a new task card for the board with an initial todo status.";
  }

  @Override
  public AppSchema inputSchema() {
    return AppSchema.object()
        .property("title", AppSchema.string().description("Visible task title shown on the card."))
        .property("description", AppSchema.string().description("Optional complementary task description."))
        .required("title")
        .additionalProperties(false);
  }

  @Override
  public AppSchema outputSchema() {
    return AppSchema.object()
        .property("task", AppSchema.object()
            .property("id", AppSchema.string())
            .property("title", AppSchema.string())
            .property("description", AppSchema.string())
            .property("status", AppSchema.string().enumValues("todo", "doing", "done"))
            .property("createdAt", AppSchema.string())
            .property("updatedAt", AppSchema.string())
            .required("id", "title", "status", "createdAt", "updatedAt")
            .additionalProperties(false))
        .required("task")
        .additionalProperties(false);
  }

  @Override
  public void validate(TaskCreateInput input) {
    if (input == null) {
      throw new IllegalArgumentException("input must be an object");
    }

    if (input.title() == null || input.title().trim().isEmpty()) {
      throw new IllegalArgumentException("title must not be empty");
    }
  }

  @Override
  public List<String> invariants() {
    return List.of(
        "Every new task starts with status todo.",
        "The backend is the source of truth for task id and timestamps.",
        "createdAt and updatedAt are equal on first creation."
    );
  }

  @Override
  public List<DomainExample<TaskCreateInput, TaskCreateOutput>> examples() {
    DomainExample<TaskCreateInput, TaskCreateOutput> titleOnly = new DomainExample<>();
    titleOnly.name = "title_only";
    titleOnly.description = "Create a task with only the required title.";
    titleOnly.input = new TaskCreateInput("Ship the Java example", null);
    titleOnly.output = new TaskCreateOutput(
        new Task(
            "task_001",
            "Ship the Java example",
            null,
            "todo",
            "2026-03-18T12:00:00.000Z",
            "2026-03-18T12:00:00.000Z"
        )
    );

    DomainExample<TaskCreateInput, TaskCreateOutput> titleAndDescription = new DomainExample<>();
    titleAndDescription.name = "title_and_description";
    titleAndDescription.description = "Create a task with an optional description.";
    titleAndDescription.input = new TaskCreateInput(
        "Prepare release notes",
        "Summarize the scope of the first Java APP example."
    );
    titleAndDescription.output = new TaskCreateOutput(
        new Task(
            "task_002",
            "Prepare release notes",
            "Summarize the scope of the first Java APP example.",
            "todo",
            "2026-03-18T12:10:00.000Z",
            "2026-03-18T12:10:00.000Z"
        )
    );

    return List.of(titleOnly, titleAndDescription);
  }

  @Override
  public void test() throws Exception {
    if (!"task_create".equals(caseName())) {
      throw new IllegalStateException("test: caseName must be task_create");
    }

    validate(new TaskCreateInput("Valid task", "Optional text"));

    boolean rejected = false;
    try {
      validate(new TaskCreateInput("   ", null));
    } catch (IllegalArgumentException ignored) {
      rejected = true;
    }

    if (!rejected) {
      throw new IllegalStateException("test: validate must reject blank title");
    }

    rejected = false;
    try {
      TaskCreateInput.fromMap(Map.of("title", "Bad", "status", "todo"));
    } catch (IllegalArgumentException ignored) {
      rejected = true;
    }

    if (!rejected) {
      throw new IllegalStateException("test: fromMap must reject forbidden fields");
    }

    for (DomainExample<TaskCreateInput, TaskCreateOutput> example : examples()) {
      validate(example.input);
      if (!"todo".equals(example.output.task().status())) {
        throw new IllegalStateException("test: example output must start in todo");
      }
    }
  }

  private static String requiredString(Map<String, Object> value, String field, String source) {
    Object current = value.get(field);
    if (!(current instanceof String stringValue) || stringValue.trim().isEmpty()) {
      throw new IllegalArgumentException(source + "." + field + " must be a non-empty string");
    }

    return stringValue;
  }
}

File: examples/python/cases/tasks/task_create/task_create_domain_case.py
# ==========================================================================
# APP v1.1.4
# cases/tasks/task_create/task_create_domain_case.py
# --------------------------------------------------------------------------
# Domain surface for the task_create Case.
#
# Responsibility:
# - define task types and shared contracts
# - validate input (pure, no I/O)
# - expose invariants, examples, and schemas
# - no infrastructure access
# ==========================================================================

from __future__ import annotations

from typing import Literal, TypedDict

from core.domain_case import AppSchema, BaseDomainCase, DomainExample
from core.shared.app_structural_contracts import AppCaseError

# ==========================================================================
# Shared types
# --------------------------------------------------------------------------
# Defined here so each surface file can import what it needs without
# cross-surface coupling. Cases are self-contained.
# ==========================================================================

TaskStatus = Literal["todo", "doing", "done"]


class Task(TypedDict):
    id: str
    title: str
    status: TaskStatus
    createdAt: str
    updatedAt: str


class TaskCreateInput(TypedDict, total=False):
    title: str  # required


class TaskCreateOutput(TypedDict):
    task: Task


# ==========================================================================
# TaskCreateDomain
# ==========================================================================

class TaskCreateDomain(BaseDomainCase[TaskCreateInput, TaskCreateOutput]):
    """Domain surface for the task_create Case."""

    def case_name(self) -> str:
        return "task_create"

    def description(self) -> str:
        return "Creates a new task card for the board with an initial todo status."

    def input_schema(self) -> AppSchema:
        return {
            "type": "object",
            "properties": {
                "title": {
                    "type": "string",
                    "description": "Visible task title shown on the card.",
                },
            },
            "required": ["title"],
            "additionalProperties": False,
        }

    def output_schema(self) -> AppSchema:
        return {
            "type": "object",
            "properties": {
                "task": {
                    "type": "object",
                    "properties": {
                        "id": {"type": "string"},
                        "title": {"type": "string"},
                        "status": {
                            "type": "string",
                            "enum": ["todo", "doing", "done"],
                        },
                        "createdAt": {"type": "string"},
                        "updatedAt": {"type": "string"},
                    },
                    "required": ["id", "title", "status", "createdAt", "updatedAt"],
                    "additionalProperties": False,
                },
            },
            "required": ["task"],
            "additionalProperties": False,
        }

    def validate(self, input: TaskCreateInput) -> None:
        """
        Pure input validation. No I/O.

        Raises AppCaseError("VALIDATION_FAILED", ...) on any violation.
        """
        if not isinstance(input, dict):
            raise AppCaseError("VALIDATION_FAILED", "input must be an object")

        title = input.get("title")
        if title is None or not isinstance(title, str):
            raise AppCaseError(
                "VALIDATION_FAILED",
                "title is required and must be a string",
            )

        if title.strip() == "":
            raise AppCaseError("VALIDATION_FAILED", "title must not be empty")

        # Forbidden fields — the backend owns id, status, and timestamps
        forbidden = ["id", "status", "createdAt", "updatedAt"]
        for field in forbidden:
            if field in input:
                raise AppCaseError(
                    "VALIDATION_FAILED",
                    f"{field} must not be provided by the caller",
                )

    def invariants(self) -> list[str]:
        return [
            "Every new task starts with status todo.",
            "The backend is the source of truth for task id and timestamps.",
            "createdAt and updatedAt are equal on first creation.",
        ]

    def examples(self) -> list[DomainExample[TaskCreateInput, TaskCreateOutput]]:
        return [
            DomainExample(
                name="title_only",
                description="Create a task with only the required title.",
                input={"title": "Ship the Python example"},
                output={
                    "task": {
                        "id": "task_001",
                        "title": "Ship the Python example",
                        "status": "todo",
                        "createdAt": "2026-03-18T12:00:00.000000+00:00",
                        "updatedAt": "2026-03-18T12:00:00.000000+00:00",
                    }
                },
            ),
            DomainExample(
                name="title_and_description",
                description="Create a task with a descriptive title.",
                input={"title": "Prepare release notes"},
                output={
                    "task": {
                        "id": "task_002",
                        "title": "Prepare release notes",
                        "status": "todo",
                        "createdAt": "2026-03-18T12:10:00.000000+00:00",
                        "updatedAt": "2026-03-18T12:10:00.000000+00:00",
                    }
                },
            ),
        ]

    def test(self) -> None:
        """Self-contained domain surface test."""

        # --- Structural validation ---
        definition = self.definition()

        if definition["case_name"] != "task_create":
            raise AssertionError("test: case_name must be task_create")

        input_schema = definition["input_schema"]
        if "title" not in (input_schema.get("required") or []):
            raise AssertionError("test: input_schema must require title")

        if input_schema.get("additionalProperties") is not False:
            raise AssertionError("test: input_schema must set additionalProperties false")

        # --- Valid input accepted ---
        self.validate({"title": "Valid task"})

        # --- Blank title rejected ---
        threw = False
        try:
            self.validate({"title": "   "})
        except AppCaseError:
            threw = True
        if not threw:
            raise AssertionError("test: validate must reject blank title")

        # --- Missing title rejected ---
        threw = False
        try:
            self.validate({})  # type: ignore[arg-type]
        except AppCaseError:
            threw = True
        if not threw:
            raise AssertionError("test: validate must reject missing title")

        # --- Forbidden fields rejected ---
        threw = False
        try:
            self.validate({"title": "Bad task", "status": "todo"})  # type: ignore[typeddict-unknown-key]
        except AppCaseError:
            threw = True
        if not threw:
            raise AssertionError("test: validate must reject forbidden field status")

        threw = False
        try:
            self.validate({"title": "Bad task", "id": "x"})  # type: ignore[typeddict-unknown-key]
        except AppCaseError:
            threw = True
        if not threw:
            raise AssertionError("test: validate must reject forbidden field id")

        # --- Examples validation ---
        examples = self.examples()
        if not examples:
            raise AssertionError("test: examples must be defined")

        for example in examples:
            self.validate(example.input)
            if example.output is None:
                raise AssertionError("test: example output must be present")
            if example.output["task"]["status"] != "todo":
                raise AssertionError("test: example output must start in todo")
            if example.output["task"]["createdAt"] != example.output["task"]["updatedAt"]:
                raise AssertionError("test: example createdAt and updatedAt must be equal on creation")
